### PR TITLE
Support additional API on Python 3 bindings

### PR DESCRIPTION
**Pull request by elicn**
Highlights:
 - Redesign certain architecture-specific aspects to better leverage OOP
   - Code de-duplication
   - Easier to extend or support new architectures
 - Now fully support `reg_write_batch` and `reg_read_batch` API
 - Introduced coprocessor register accessors to AArch32 and AArch64, named `cpr_read` and `cpr_write`
 - 1 ARM regression test fixed
 - More comments, documentation and annotations

### CODE REVIEW

File: bindings/python/unicorn/unicorn_py3/arch/arm.py [hunk 1]
```diff
@@ -40,46 +40,71 @@ class UcAArch32(Uc):
...
+            is_64   : indicates whether this is a 64-bit register
 
-        return self._reg_read(reg_id, reg_cls)
+        Returns: value of coprocessor register
+        """
 
-    def reg_write(self, reg_id: int, value) -> None:
-        # select register class for special cases
-        reg_cls = UcAArch32.__select_reg_class(reg_id)
+        assert check_maxbits(coproc, 4)
```
Comment (wtdcode) [inline bindings/python/unicorn/unicorn_py3/arch/arm.py:77]: Why not make `cpr_read/write` return `uc_err` and avoid `assert` here by returning `UC_ERR_ARG`? Terminating the whole process seems too aggressive.

File: bindings/python/unicorn/unicorn_py3/unicorn.py [hunk 2]
```diff
@@ -1227,11 +1359,17 @@ def _do_reg_read_batch(self, reglist, vallist, count) -> int:
 
         return uclib.uc_context_reg_read_batch(self._context, reglist, vallist, count)
 
+    def _do_reg_write_batch(self, reglist, vallist, count) -> int:
```
Comment (wtdcode) [inline bindings/python/unicorn/unicorn_py3/unicorn.py:1362]: Duplicate definition of line 762?

File: bindings/python/unicorn/unicorn_py3/arch/arm.py (hunk 1 above)
Comment (elicn) [reply bindings/python/unicorn/unicorn_py3/arch/arm.py:77]: Generally speaking, exceptions are thrown when the arguments normally make sense but there is a specific problem this time (e.g. opening a file which happens to be inaccessible, etc.). In these cases the user may catch the exception and choose how to handle it.

Here the assertions come to block an invalid usage with incorrect arguments: it is not a call that may sometimes work and sometimes not -- rather it is invalid. There is no way to "handle" such case other than re-writing that piece of code.

File: bindings/python/unicorn/unicorn_py3/unicorn.py (hunk 2 above)
Comment (elicn) [reply bindings/python/unicorn/unicorn_py3/unicorn.py:1362]: They belong to two different classes (engine and context). The underlying methods are using the same names to allow seamless OOP.

Comment (wtdcode) [reply bindings/python/unicorn/unicorn_py3/unicorn.py:1362]: Okay I see

File: bindings/python/unicorn/unicorn_py3/arch/intel.py [hunk 3]
```diff
@@ -127,67 +129,42 @@ def __hook_insn_cpuid_cb(uc: Uc, key: int) -> int:
...
 
-            super().reg_write(reg_id, value)
-        else:
-            self._reg_write(reg_id, reg_cls, value)
-
-    def msr_read(self, msr_id: int) -> int:
-        return self._reg_read(const.UC_X86_REG_MSR, UcRegMSR, msr_id)
+        return self.reg_read(const.UC_X86_REG_MSR, msr_id)
 
     def msr_write(self, msr_id: int, value: int) -> None:
```
[the remark this answers is not in the document]
Comment (elicn) [reply bindings/python/unicorn/unicorn_py3/arch/intel.py:159]: I am not following on this one. What should raise an exception? It marks this line:
```python
return self.reg_read(const.UC_X86_REG_MSR, msr_id)
```

File: bindings/python/unicorn/unicorn_py3/arch/arm.py (hunk 1 above)
Comment (wtdcode) [reply bindings/python/unicorn/unicorn_py3/arch/arm.py:77]: Okay, I also forget `AssertionError` can be caught so let it go.

**Pull request merged**